// openclaw/dist/plugin-sdk/memory-core-host-runtime-cli.js
import { i as formatErrorMessage } from "../errors-QN8rySzW.js";
import { t as formatDocsLink } from "../links-dQIIPEtq.js";
import { n as isRich, r as theme, t as colorize } from "../theme-CVJvORNs.js";
import { g as shortenHomePath, h as shortenHomeInString } from "../utils-D5swhEXt.js";
import { n as defaultRuntime } from "../runtime-bzt9CHmD.js";
import { S as setVerbose, b as isVerbose } from "../logger-BVNXvwCE.js";
import { t as resolveCommandSecretRefsViaGateway } from "../command-secret-gateway-CSCvKTM0.js";
import { i as withProgressTotals, r as withProgress } from "../progress-BUoAGuhg.js";
import { r as withManager } from "../cli-utils-BLmbV6RC.js";
import { t as formatHelpExamples } from "../help-format-y64qVlFX.js";
import "../memory-core-host-runtime-cli-BzOsobfD.js";
export { colorize, defaultRuntime, formatDocsLink, formatErrorMessage, formatHelpExamples, isRich, isVerbose, resolveCommandSecretRefsViaGateway, setVerbose, shortenHomeInString, shortenHomePath, theme, withManager, withProgress, withProgressTotals };
