// tokenjuice/dist/hosts/shared/hook-doctor.js
import { doctorAiderConvention } from "../aider/index.js";
import { doctorAvanteInstructions } from "../avante/index.js";
import { doctorClaudeCodeHook } from "../claude-code/index.js";
import { doctorClineHook } from "../cline/index.js";
import { doctorCodeBuddyHook } from "../codebuddy/index.js";
import { doctorContinueRule } from "../continue/index.js";
import { doctorCodexHook } from "../codex/index.js";
import { doctorCopilotCliHook } from "../copilot-cli/index.js";
import { doctorCursorHook } from "../cursor/index.js";
import { doctorDroidHook } from "../droid/index.js";
import { doctorGeminiCliHook } from "../gemini-cli/index.js";
import { doctorJunieInstructions } from "../junie/index.js";
import { doctorOpenHandsHook } from "../openhands/index.js";
import { doctorPiExtension } from "../pi/index.js";
import { doctorVscodeCopilotHook } from "../vscode-copilot/index.js";
import { doctorZedInstructions } from "../zed/index.js";
const hookDoctorIntegrationDoctors = {
    aider: () => doctorAiderConvention(),
    avante: () => doctorAvanteInstructions(),
    codex: (options) => doctorCodexHook(undefined, options),
    "claude-code": (options) => doctorClaudeCodeHook(undefined, getHookCommandOptions(options)),
    cline: (options) => doctorClineHook(undefined, getHookCommandOptions(options)),
    codebuddy: (options) => doctorCodeBuddyHook(undefined, getHookCommandOptions(options)),
    continue: () => doctorContinueRule(),
    cursor: (options) => doctorCursorHook(undefined, getHookCommandOptions(options)),
    droid: (options) => doctorDroidHook(undefined, getHookCommandOptions(options)),
    "gemini-cli": (options) => doctorGeminiCliHook(undefined, getHookCommandOptions(options)),
    junie: () => doctorJunieInstructions(),
    openhands: (options) => doctorOpenHandsHook(undefined, getHookCommandOptions(options)),
    pi: () => doctorPiExtension(),
    "vscode-copilot": (options) => doctorVscodeCopilotHook(undefined, getHookCommandOptions(options)),
    zed: () => doctorZedInstructions(),
    "copilot-cli": (options) => doctorCopilotCliHook(undefined, getHookCommandOptions(options)),
};
export function getAvailableHookIntegrationNames() {
    return Object.keys(hookDoctorIntegrationDoctors);
}
function mergeStatus(left, right) {
    if (left === "broken" || right === "broken") {
        return "broken";
    }
    if (left === "warn" || right === "warn") {
        return "warn";
    }
    if (left === "ok" || right === "ok") {
        return "ok";
    }
    if (left === "disabled" || right === "disabled") {
        return "disabled";
    }
    return "ok";
}
function mergeStatuses(statuses) {
    return statuses.reduce(mergeStatus, "disabled");
}
function getHookCommandOptions(options) {
    return {
        ...(typeof options.local === "boolean" ? { local: options.local } : {}),
        ...(typeof options.binaryPath === "string" ? { binaryPath: options.binaryPath } : {}),
        ...(typeof options.nodePath === "string" ? { nodePath: options.nodePath } : {}),
    };
}
function hasDetectedCommand(report) {
    return "detectedCommand" in report && typeof report.detectedCommand === "string" && report.detectedCommand.length > 0;
}
export function isInstalledHookIntegration(report) {
    if (hasDetectedCommand(report)) {
        return true;
    }
    // Command-backed hosts can warn about missing config even when tokenjuice is
    // not installed. Instruction/extension hosts have no expected command, so a
    // non-disabled status means their tokenjuice artifact exists.
    return report.status !== "disabled" && !("expectedCommand" in report);
}
export function getInstalledHookIntegrations(report) {
    return Object.entries(report.integrations)
        .filter(([, integrationReport]) => isInstalledHookIntegration(integrationReport));
}
export async function doctorInstalledHooks(options = {}) {
    const integrationEntries = await Promise.all(Object.entries(hookDoctorIntegrationDoctors).map(async ([name, doctor]) => [name, await doctor(options)]));
    const integrations = Object.fromEntries(integrationEntries);
    const installedIntegrations = getInstalledHookIntegrations({ status: "disabled", integrations });
    return {
        status: mergeStatuses(installedIntegrations.map(([, integrationReport]) => integrationReport.status)),
        integrations,
    };
}
