// tokenjuice/dist/hosts/shared/hooks-json-file.js
import { mkdir, readFile, readdir, rename, writeFile } from "node:fs/promises";
import { join } from "node:path";
export function isRecord(value) {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}
export function sanitizeCopilotHooksConfig(raw) {
    if (!isRecord(raw)) {
        return { version: 1, hooks: {} };
    }
    return {
        ...raw,
        version: typeof raw.version === "number" ? raw.version : 1,
        hooks: isRecord(raw.hooks) ? { ...raw.hooks } : {},
    };
}
export async function readCopilotHooksConfig(hooksPath, hostLabel) {
    try {
        const rawText = await readFile(hooksPath, "utf8");
        const parsed = JSON.parse(rawText);
        return { config: sanitizeCopilotHooksConfig(parsed), exists: true };
    }
    catch (error) {
        if (error.code === "ENOENT") {
            return { config: { version: 1, hooks: {} }, exists: false };
        }
        throw new Error(`failed to read ${hostLabel} hooks from ${hooksPath}: ${error instanceof Error ? error.message : String(error)}`);
    }
}
export async function loadCopilotHooksConfigWithBackup(hooksPath, hostLabel) {
    try {
        const rawText = await readFile(hooksPath, "utf8");
        const parsed = JSON.parse(rawText);
        const config = sanitizeCopilotHooksConfig(parsed);
        const backupPath = `${hooksPath}.bak`;
        await writeFile(backupPath, rawText, "utf8");
        return { config, backupPath };
    }
    catch (error) {
        if (error.code === "ENOENT") {
            return { config: { version: 1, hooks: {} } };
        }
        throw new Error(`failed to load ${hostLabel} hooks from ${hooksPath}: ${error instanceof Error ? error.message : String(error)}`);
    }
}
export async function writeCopilotHooksConfigAtomic(hooksPath, config) {
    const tempPath = `${hooksPath}.tmp`;
    await writeFile(tempPath, `${JSON.stringify(config, null, 2)}\n`, "utf8");
    await rename(tempPath, hooksPath);
}
export async function ensureHooksDir(hooksDir) {
    await mkdir(hooksDir, { recursive: true });
}
// Scan every sibling `*.json` in `hooksDir`, skipping the canonical
// install path, and return those that contain a tokenjuice entry
// matching `isTokenjuiceCommand`. Used by doctor to surface stray
// installs that would otherwise be silently double-loaded by the
// host's hook scanner.
export async function findStrayTokenjuiceHookFiles(hooksDir, canonicalPath, isTokenjuiceCommand) {
    let entries;
    try {
        entries = await readdir(hooksDir);
    }
    catch {
        return [];
    }
    const stray = [];
    for (const entry of entries) {
        if (!entry.endsWith(".json")) {
            continue;
        }
        const candidate = join(hooksDir, entry);
        if (candidate === canonicalPath) {
            continue;
        }
        try {
            const rawText = await readFile(candidate, "utf8");
            const parsed = JSON.parse(rawText);
            const config = sanitizeCopilotHooksConfig(parsed);
            if (hasTokenjuiceCommand(config, isTokenjuiceCommand)) {
                stray.push(candidate);
            }
        }
        catch {
            // ignore unreadable/invalid files
        }
    }
    return stray;
}
function hasTokenjuiceCommand(config, isTokenjuiceCommand) {
    for (const bucket of Object.values(config.hooks)) {
        if (!Array.isArray(bucket)) {
            continue;
        }
        for (const entry of bucket) {
            if (isRecord(entry)
                && typeof entry.command === "string"
                && isTokenjuiceCommand(entry.command)) {
                return true;
            }
        }
    }
    return false;
}
