// tokenjuice/dist/hosts/shared/hook-command-doctor.js
import { findMissingHookCommandPaths } from "./host-command.js";
export async function buildHookCommandDoctorFields(options) {
    if (!options.detectedCommand) {
        return {
            status: "disabled",
            issues: [options.disabledIssue],
            advisories: [options.advisory],
            fixCommand: options.fixCommand,
            expectedCommand: options.expectedCommand,
            checkedPaths: [],
            missingPaths: [],
        };
    }
    const missingPaths = await findMissingHookCommandPaths(options.detectedCommand);
    const issues = [];
    if (options.detectedCommand !== options.expectedCommand) {
        issues.push(`configured ${options.hostLabel} hook command does not match the current recommended command`);
    }
    if (missingPaths.length > 0) {
        issues.push(`configured ${options.hostLabel} hook points at missing path${missingPaths.length === 1 ? "" : "s"}`);
    }
    return {
        status: issues.length > 0 ? "broken" : "ok",
        issues,
        advisories: [options.advisory],
        fixCommand: options.fixCommand,
        expectedCommand: options.expectedCommand,
        detectedCommand: options.detectedCommand,
        checkedPaths: [],
        missingPaths,
    };
}
