// tokenjuice/dist/hosts/shared/marker-instructions.js
import { rm, writeFile } from "node:fs/promises";
import { readInstructionFile, writeInstructionFile } from "./instruction-file.js";
function escapeRegExp(value) {
    return value.replace(/[.*+?^${}()|[\]\\]/gu, "\\$&");
}
function buildBlockPattern(config) {
    return new RegExp(`\\n?${escapeRegExp(config.beginMarker)}[\\s\\S]*?${escapeRegExp(config.endMarker)}\\n?`, "gu");
}
export function inspectMarkerDelimitedBlock(text, config) {
    return {
        hasBegin: text.includes(config.beginMarker),
        hasEnd: text.includes(config.endMarker),
        completeBlockCount: [...text.matchAll(buildBlockPattern(config))].length,
    };
}
export function collectMarkerDelimitedBlockIssues(state, options) {
    if (state.hasBegin && !state.hasEnd) {
        return [`configured ${options.configuredLabel} have a tokenjuice start marker without an end marker`];
    }
    if (!state.hasBegin && state.hasEnd) {
        return [`configured ${options.configuredLabel} have a tokenjuice end marker without a start marker`];
    }
    if (state.completeBlockCount !== 1) {
        return [`configured ${options.configuredLabel} have multiple tokenjuice blocks; run ${options.repairCommand} to repair`];
    }
    return [];
}
export function removeMarkerDelimitedBlock(text, config) {
    const pattern = buildBlockPattern(config);
    if (!pattern.test(text)) {
        return { text, removed: false };
    }
    return {
        text: text.replace(buildBlockPattern(config), "\n").replace(/\n{3,}/gu, "\n\n").trim(),
        removed: true,
    };
}
export function upsertMarkerDelimitedBlock(text, config) {
    const withoutBlock = removeMarkerDelimitedBlock(text, config).text.trim();
    if (!withoutBlock) {
        return `${config.block}\n`;
    }
    return `${withoutBlock}\n\n${config.block}\n`;
}
export async function installMarkerDelimitedBlock(filePath, config) {
    const existing = await readInstructionFile(filePath);
    const result = await writeInstructionFile(filePath, upsertMarkerDelimitedBlock(existing.text, config));
    return {
        filePath: result.filePath,
        ...(result.backupPath ? { backupPath: result.backupPath } : {}),
    };
}
export async function uninstallMarkerDelimitedBlock(filePath, config) {
    const existing = await readInstructionFile(filePath);
    if (!existing.exists) {
        return { filePath, removed: false };
    }
    const removed = removeMarkerDelimitedBlock(existing.text, config);
    if (!removed.removed) {
        return { filePath, removed: false };
    }
    if (removed.text.trim()) {
        await writeFile(filePath, `${removed.text.trim()}\n`, "utf8");
    }
    else {
        await rm(filePath, { force: true });
    }
    return { filePath, removed: true };
}
