// tokenjuice/dist/hosts/avante/index.js
import { join } from "node:path";
import { collectMarkerDelimitedBlockIssues, inspectMarkerDelimitedBlock, installMarkerDelimitedBlock, uninstallMarkerDelimitedBlock, } from "../shared/marker-instructions.js";
import { buildTokenjuiceGuidanceBullets } from "../shared/instruction-guidance.js";
import { buildInstructionDoctorReportFields, instructionDoctorStatusFromIssues, } from "../shared/instruction-doctor.js";
import { readInstructionFile } from "../shared/instruction-file.js";
const TOKENJUICE_AVANTE_FIX_COMMAND = "tokenjuice install avante";
const TOKENJUICE_AVANTE_BEGIN = "<!-- tokenjuice:begin -->";
const TOKENJUICE_AVANTE_END = "<!-- tokenjuice:end -->";
const TOKENJUICE_AVANTE_ADVISORY = "Avante support is beta and instruction-based; it guides command usage but does not intercept tool output.";
function getProjectDir(options = {}) {
    return options.projectDir || process.env.AVANTE_PROJECT_DIR || process.cwd();
}
function getDefaultInstructionsPath(options = {}) {
    return join(getProjectDir(options), "avante.md");
}
const TOKENJUICE_AVANTE_BLOCK = [
    TOKENJUICE_AVANTE_BEGIN,
    "## tokenjuice terminal output compaction",
    "",
    ...buildTokenjuiceGuidanceBullets(),
    TOKENJUICE_AVANTE_END,
].join("\n");
const TOKENJUICE_AVANTE_BLOCK_CONFIG = {
    beginMarker: TOKENJUICE_AVANTE_BEGIN,
    endMarker: TOKENJUICE_AVANTE_END,
    block: TOKENJUICE_AVANTE_BLOCK,
};
export async function installAvanteInstructions(instructionsPath, options = {}) {
    const resolvedInstructionsPath = instructionsPath ?? getDefaultInstructionsPath(options);
    const result = await installMarkerDelimitedBlock(resolvedInstructionsPath, TOKENJUICE_AVANTE_BLOCK_CONFIG);
    return {
        instructionsPath: result.filePath,
        ...(result.backupPath ? { backupPath: result.backupPath } : {}),
    };
}
export async function uninstallAvanteInstructions(instructionsPath = getDefaultInstructionsPath()) {
    const result = await uninstallMarkerDelimitedBlock(instructionsPath, TOKENJUICE_AVANTE_BLOCK_CONFIG);
    return { instructionsPath: result.filePath, removed: result.removed };
}
export async function doctorAvanteInstructions(instructionsPath, options = {}) {
    const resolvedInstructionsPath = instructionsPath ?? getDefaultInstructionsPath(options);
    const existing = await readInstructionFile(resolvedInstructionsPath);
    const markerState = inspectMarkerDelimitedBlock(existing.text, TOKENJUICE_AVANTE_BLOCK_CONFIG);
    if (!existing.exists || (!markerState.hasBegin && !markerState.hasEnd)) {
        return {
            instructionsPath: resolvedInstructionsPath,
            ...buildInstructionDoctorReportFields({
                status: "disabled",
                issues: ["tokenjuice Avante instructions are not installed"],
                advisory: TOKENJUICE_AVANTE_ADVISORY,
                fixCommand: TOKENJUICE_AVANTE_FIX_COMMAND,
            }),
        };
    }
    const issues = collectMarkerDelimitedBlockIssues(markerState, {
        configuredLabel: "Avante instructions",
        repairCommand: TOKENJUICE_AVANTE_FIX_COMMAND,
    });
    return {
        instructionsPath: resolvedInstructionsPath,
        ...buildInstructionDoctorReportFields({
            status: instructionDoctorStatusFromIssues(issues),
            issues,
            advisory: TOKENJUICE_AVANTE_ADVISORY,
            fixCommand: TOKENJUICE_AVANTE_FIX_COMMAND,
        }),
    };
}
