// openclaw/dist/extensions/bonjour/index.js
import { t as definePluginEntry } from "../../plugin-entry-CJ7dbRiF.js";
//#region extensions/bonjour/index.ts
function formatBonjourInstanceName(displayName) {
	const trimmed = displayName.trim();
	if (!trimmed) return "OpenClaw";
	if (/openclaw/i.test(trimmed)) return trimmed;
	return `${trimmed} (OpenClaw)`;
}
var bonjour_default = definePluginEntry({
	id: "bonjour",
	name: "Bonjour Gateway Discovery",
	description: "Advertise the local OpenClaw gateway over Bonjour/mDNS.",
	register(api) {
		api.registerGatewayDiscoveryService({
			id: "bonjour",
			advertise: async (ctx) => {
				const [{ startGatewayBonjourAdvertiser }, { registerUncaughtExceptionHandler, registerUnhandledRejectionHandler }] = await Promise.all([import("../../advertiser-RTXe0iPg.js"), import("../../plugin-sdk/runtime.js")]);
				return { stop: (await startGatewayBonjourAdvertiser({
					instanceName: formatBonjourInstanceName(ctx.machineDisplayName),
					gatewayPort: ctx.gatewayPort,
					gatewayTlsEnabled: ctx.gatewayTlsEnabled,
					gatewayTlsFingerprintSha256: ctx.gatewayTlsFingerprintSha256,
					canvasPort: ctx.canvasPort,
					sshPort: ctx.sshPort,
					tailnetDns: ctx.tailnetDns,
					cliPath: ctx.cliPath,
					minimal: ctx.minimal
				}, {
					logger: api.logger,
					registerUncaughtExceptionHandler,
					registerUnhandledRejectionHandler
				})).stop };
			}
		});
	}
});
//#endregion
export { bonjour_default as default };
