// tokenjuice/dist/core/reduce-utils.js
import { createHash } from "node:crypto";
import { createCompactionMetadata } from "./compaction-metadata.js";
import { countTextChars, sliceTextChars } from "./text.js";
export function compactWhitespace(text) {
    return text.replace(/\s+/gu, " ").trim();
}
function shortHash(text) {
    return createHash("sha256").update(text).digest("hex").slice(0, 12);
}
export function clipMiddleWithHash(text, maxChars) {
    if (countTextChars(text) <= maxChars) {
        return { text };
    }
    const omitted = countTextChars(text) - maxChars;
    const headChars = Math.max(20, Math.floor(maxChars * 0.55));
    const tailChars = Math.max(20, maxChars - headChars);
    return {
        text: `${sliceTextChars(text, 0, headChars)} ...[${omitted} chars omitted, sha256:${shortHash(text)}]... ${sliceTextChars(text, -tailChars)}`,
        compaction: createCompactionMetadata("hashed-middle-clip"),
    };
}
export function parseJsonObjectLine(line) {
    const trimmed = line.trim();
    if (!trimmed.startsWith("{") || !trimmed.endsWith("}")) {
        return null;
    }
    try {
        const parsed = JSON.parse(trimmed);
        return typeof parsed === "object" && parsed !== null && !Array.isArray(parsed)
            ? parsed
            : null;
    }
    catch {
        return null;
    }
}
export function parseJsonValue(text) {
    const trimmed = text.trim();
    if (!(trimmed.startsWith("{") || trimmed.startsWith("["))) {
        return null;
    }
    try {
        return JSON.parse(trimmed);
    }
    catch {
        return null;
    }
}
