// tokenjuice/dist/core/rules.js
import { readdir, readFile, realpath } from "node:fs/promises";
import { homedir } from "node:os";
import { isAbsolute, join, relative, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { BUNDLED_BUILTIN_RULES } from "./builtin-rules.generated.js";
import { assertValidRule, validateRule } from "./validate-rules.js";
const ruleCache = new Map();
function builtinRulesRoot() {
    return resolve(fileURLToPath(new URL("../rules", import.meta.url)));
}
function mergeRegexFlags(flags) {
    return [...new Set(`u${flags ?? ""}`.split(""))].join("");
}
function compileRule(descriptor) {
    return {
        rule: descriptor.rule,
        source: descriptor.source,
        path: descriptor.path,
        compiled: {
            skipPatterns: (descriptor.rule.filters?.skipPatterns ?? []).map((pattern) => new RegExp(pattern, "u")),
            keepPatterns: (descriptor.rule.filters?.keepPatterns ?? []).map((pattern) => new RegExp(pattern, "u")),
            counters: (descriptor.rule.counters ?? []).map((counter) => ({
                name: counter.name,
                pattern: new RegExp(counter.pattern, mergeRegexFlags(counter.flags)),
            })),
            outputMatches: (descriptor.rule.matchOutput ?? []).map((entry) => ({
                pattern: new RegExp(entry.pattern, mergeRegexFlags(entry.flags)),
                message: entry.message,
            })),
        },
    };
}
async function listRuleFiles(root) {
    const resolvedRoot = await realpath(root).catch(() => null);
    if (!resolvedRoot) {
        return [];
    }
    const rootRealPath = resolvedRoot;
    async function walk(currentDir) {
        const entries = (await readdir(currentDir, { withFileTypes: true })).sort((left, right) => left.name.localeCompare(right.name));
        const files = await Promise.all(entries.map(async (entry) => {
            const fullPath = join(currentDir, entry.name);
            if (entry.isSymbolicLink()) {
                return [];
            }
            if (entry.isDirectory()) {
                return await walk(fullPath);
            }
            if (!entry.isFile()
                || !entry.name.endsWith(".json")
                || entry.name.endsWith(".schema.json")
                || entry.name.endsWith(".fixture.json")) {
                return [];
            }
            const realFilePath = await realpath(fullPath).catch(() => null);
            if (!realFilePath) {
                return [];
            }
            const relativePath = relative(rootRealPath, realFilePath);
            if (relativePath.startsWith("..") || isAbsolute(relativePath)) {
                return [];
            }
            return [realFilePath];
        }));
        return files.flat();
    }
    try {
        return await walk(root);
    }
    catch {
        return [];
    }
}
function userRulesRoot(customDir) {
    return customDir ?? join(homedir(), ".config", "tokenjuice", "rules");
}
function projectRulesRoot(cwd, customDir) {
    return customDir ?? join(cwd ?? process.cwd(), ".tokenjuice", "rules");
}
function sortRules(rules) {
    return [...rules].sort((left, right) => {
        if (left.rule.id === "generic/fallback") {
            return 1;
        }
        if (right.rule.id === "generic/fallback") {
            return -1;
        }
        return left.rule.id.localeCompare(right.rule.id);
    });
}
async function loadRuleDescriptorsFromRoot(root, source) {
    const files = await listRuleFiles(root);
    return await Promise.all(files.map(async (fullPath) => {
        const parsed = JSON.parse(await readFile(fullPath, "utf8"));
        assertValidRule(parsed);
        return {
            source,
            path: fullPath,
            relativePath: relative(root, fullPath),
            rule: parsed,
        };
    }));
}
function loadBundledBuiltinRuleDescriptors() {
    return BUNDLED_BUILTIN_RULES.map((rule) => {
        assertValidRule(rule);
        return {
            source: "builtin",
            path: `bundled:${rule.id}`,
            relativePath: rule.id,
            rule,
        };
    });
}
function cacheKey(options) {
    return JSON.stringify({
        cwd: options.cwd ?? process.cwd(),
        includeUser: options.includeUser ?? true,
        includeProject: options.includeProject ?? true,
        userRulesDir: options.userRulesDir ?? null,
        projectRulesDir: options.projectRulesDir ?? null,
    });
}
function overlayRules(descriptors) {
    const byId = new Map();
    for (const descriptor of descriptors) {
        byId.set(descriptor.rule.id, descriptor);
    }
    return sortRules([...byId.values()].map((descriptor) => compileRule(descriptor)));
}
export async function loadRules(options = {}) {
    const key = cacheKey(options);
    const cached = ruleCache.get(key);
    if (cached) {
        return cached;
    }
    const descriptors = [];
    const builtinDescriptors = await loadRuleDescriptorsFromRoot(builtinRulesRoot(), "builtin");
    descriptors.push(...(builtinDescriptors.length > 0 ? builtinDescriptors : loadBundledBuiltinRuleDescriptors()));
    if (options.includeUser ?? true) {
        descriptors.push(...await loadRuleDescriptorsFromRoot(userRulesRoot(options.userRulesDir), "user"));
    }
    if (options.includeProject ?? true) {
        descriptors.push(...await loadRuleDescriptorsFromRoot(projectRulesRoot(options.cwd, options.projectRulesDir), "project"));
    }
    const compiled = overlayRules(descriptors);
    ruleCache.set(key, compiled);
    return compiled;
}
export async function loadBuiltinRules() {
    return await loadRules({
        includeUser: false,
        includeProject: false,
    });
}
export function clearRuleCache() {
    ruleCache.clear();
}
async function verifyRuleRoot(root, source) {
    const files = await listRuleFiles(root);
    const results = await Promise.all(files.map(async (fullPath) => {
        try {
            const raw = JSON.parse(await readFile(fullPath, "utf8"));
            const validation = validateRule(raw);
            const errors = validation.ok ? [] : [...validation.errors];
            if (validation.ok) {
                try {
                    assertValidRule(raw);
                    compileRule({
                        source,
                        path: fullPath,
                        relativePath: relative(root, fullPath),
                        rule: raw,
                    });
                }
                catch (error) {
                    errors.push(error instanceof Error ? error.message : String(error));
                }
            }
            return {
                id: validation.ok && typeof raw === "object" && raw !== null && "id" in raw && typeof raw.id === "string"
                    ? raw.id
                    : relative(root, fullPath).replace(/\.json$/u, ""),
                ok: errors.length === 0,
                source,
                path: fullPath,
                errors,
                warnings: [],
            };
        }
        catch (error) {
            return {
                id: relative(root, fullPath).replace(/\.json$/u, ""),
                ok: false,
                source,
                path: fullPath,
                errors: [error instanceof Error ? error.message : String(error)],
                warnings: [],
            };
        }
    }));
    const idGroups = new Map();
    for (const result of results) {
        const group = idGroups.get(result.id) ?? [];
        group.push(result);
        idGroups.set(result.id, group);
    }
    for (const [id, group] of idGroups) {
        if (group.length > 1) {
            for (const result of group) {
                result.errors.push(`duplicate rule id in ${source} layer: ${id}`);
                result.ok = false;
            }
        }
    }
    return results;
}
function ruleSourceRank(source) {
    switch (source) {
        case "builtin":
            return 0;
        case "user":
            return 1;
        case "project":
            return 2;
    }
}
function addShadowWarnings(results) {
    const byId = new Map();
    for (const result of results) {
        const group = byId.get(result.id) ?? [];
        group.push(result);
        byId.set(result.id, group);
    }
    for (const [id, group] of byId) {
        if (group.length < 2) {
            continue;
        }
        const ordered = [...group].sort((left, right) => ruleSourceRank(left.source) - ruleSourceRank(right.source));
        const winner = ordered[ordered.length - 1];
        const shadowed = ordered.slice(0, -1);
        winner.warnings.push(`shadows ${shadowed.map((result) => `${result.source}:${id}`).join(", ")}`);
        for (const result of shadowed) {
            result.warnings.push(`shadowed by ${winner.source}:${id}`);
        }
    }
}
export async function verifyRules(options = {}) {
    const results = [];
    results.push(...await verifyRuleRoot(builtinRulesRoot(), "builtin"));
    if (options.includeUser ?? true) {
        results.push(...await verifyRuleRoot(userRulesRoot(options.userRulesDir), "user"));
    }
    if (options.includeProject ?? true) {
        results.push(...await verifyRuleRoot(projectRulesRoot(options.cwd, options.projectRulesDir), "project"));
    }
    addShadowWarnings(results);
    return results;
}
export async function verifyBuiltinRules() {
    return await verifyRules({
        includeUser: false,
        includeProject: false,
    });
}
