// tokenjuice/dist/core/validate-rules.js
function hasNulByte(value) {
    return value.includes("\0");
}
function validateSafeString(value, path, errors, { allowEmpty = false } = {}) {
    if (typeof value !== "string") {
        errors.push(`${path} must be a string`);
        return;
    }
    if (!allowEmpty && value.length === 0) {
        errors.push(`${path} must be a non-empty string`);
    }
    if (hasNulByte(value)) {
        errors.push(`${path} must not contain NUL bytes`);
    }
}
function validatePositiveInteger(value, path, errors) {
    if (typeof value !== "number" || !Number.isInteger(value) || value < 0) {
        errors.push(`${path} must be a non-negative integer`);
    }
}
function isRecord(value) {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}
function isStringArray(value) {
    return Array.isArray(value) && value.every((entry) => typeof entry === "string");
}
function validateMatch(value, path) {
    if (!isRecord(value)) {
        return [`${path} must be an object`];
    }
    const errors = [];
    if ("toolNames" in value && !isStringArray(value.toolNames)) {
        errors.push(`${path}.toolNames must be an array of strings`);
    }
    if ("argv0" in value && !isStringArray(value.argv0)) {
        errors.push(`${path}.argv0 must be an array of strings`);
    }
    if ("gitSubcommands" in value && !isStringArray(value.gitSubcommands)) {
        errors.push(`${path}.gitSubcommands must be an array of strings`);
    }
    if ("commandIncludes" in value && !isStringArray(value.commandIncludes)) {
        errors.push(`${path}.commandIncludes must be an array of strings`);
    }
    if ("commandIncludesAny" in value && !isStringArray(value.commandIncludesAny)) {
        errors.push(`${path}.commandIncludesAny must be an array of strings`);
    }
    if ("argvIncludes" in value) {
        if (!Array.isArray(value.argvIncludes) || !value.argvIncludes.every(isStringArray)) {
            errors.push(`${path}.argvIncludes must be an array of string arrays`);
        }
    }
    if ("argvIncludesAny" in value) {
        if (!Array.isArray(value.argvIncludesAny) || !value.argvIncludesAny.every(isStringArray)) {
            errors.push(`${path}.argvIncludesAny must be an array of string arrays`);
        }
    }
    return errors;
}
function validateCounter(value, path) {
    if (!isRecord(value)) {
        return [`${path} must be an object`];
    }
    const errors = [];
    validateSafeString(value.name, `${path}.name`, errors);
    validateSafeString(value.pattern, `${path}.pattern`, errors);
    if ("flags" in value) {
        validateSafeString(value.flags, `${path}.flags`, errors, { allowEmpty: true });
    }
    return errors;
}
function validateOutputMatch(value, path) {
    if (!isRecord(value)) {
        return [`${path} must be an object`];
    }
    const errors = [];
    validateSafeString(value.pattern, `${path}.pattern`, errors);
    validateSafeString(value.message, `${path}.message`, errors);
    if ("flags" in value) {
        validateSafeString(value.flags, `${path}.flags`, errors, { allowEmpty: true });
    }
    return errors;
}
function validateOptionalStringArrayObject(value, path, keys) {
    if (!isRecord(value)) {
        return [`${path} must be an object`];
    }
    const errors = [];
    for (const key of keys) {
        if (key in value && !isStringArray(value[key])) {
            errors.push(`${path}.${key} must be an array of strings`);
        }
    }
    return errors;
}
function validateOptionalNumberObject(value, path, keys) {
    if (!isRecord(value)) {
        return [`${path} must be an object`];
    }
    const errors = [];
    for (const key of keys) {
        if (key in value) {
            validatePositiveInteger(value[key], `${path}.${key}`, errors);
        }
    }
    return errors;
}
function validateOptionalBooleanObject(value, path, keys) {
    if (!isRecord(value)) {
        return [`${path} must be an object`];
    }
    const errors = [];
    for (const key of keys) {
        if (key in value && typeof value[key] !== "boolean") {
            errors.push(`${path}.${key} must be a boolean`);
        }
    }
    return errors;
}
export function validateRule(raw) {
    if (!isRecord(raw)) {
        return {
            ok: false,
            errors: ["rule must be an object"],
        };
    }
    const errors = [];
    validateSafeString(raw.id, "id", errors);
    validateSafeString(raw.family, "family", errors);
    if ("description" in raw) {
        validateSafeString(raw.description, "description", errors, { allowEmpty: true });
    }
    if ("onEmpty" in raw) {
        validateSafeString(raw.onEmpty, "onEmpty", errors);
    }
    if ("counterSource" in raw && raw.counterSource !== "postKeep" && raw.counterSource !== "preKeep") {
        errors.push("counterSource must be one of: postKeep, preKeep");
    }
    if ("priority" in raw && (typeof raw.priority !== "number" || !Number.isFinite(raw.priority))) {
        errors.push("priority must be a finite number");
    }
    if (!("match" in raw)) {
        errors.push("match is required");
    }
    else {
        errors.push(...validateMatch(raw.match, "match"));
    }
    if ("filters" in raw) {
        errors.push(...validateOptionalStringArrayObject(raw.filters, "filters", ["skipPatterns", "keepPatterns"]));
    }
    if ("transforms" in raw) {
        errors.push(...validateOptionalBooleanObject(raw.transforms, "transforms", [
            "stripAnsi",
            "prettyPrintJson",
            "dedupeAdjacent",
            "trimEmptyEdges",
        ]));
    }
    if ("summarize" in raw) {
        errors.push(...validateOptionalNumberObject(raw.summarize, "summarize", ["head", "tail"]));
    }
    if ("failure" in raw) {
        errors.push(...validateOptionalBooleanObject(raw.failure, "failure", ["preserveOnFailure"]));
        errors.push(...validateOptionalNumberObject(raw.failure, "failure", ["head", "tail"]));
    }
    if ("counters" in raw) {
        if (!Array.isArray(raw.counters)) {
            errors.push("counters must be an array");
        }
        else {
            raw.counters.forEach((counter, index) => {
                errors.push(...validateCounter(counter, `counters[${index}]`));
            });
        }
    }
    if ("matchOutput" in raw) {
        if (!Array.isArray(raw.matchOutput)) {
            errors.push("matchOutput must be an array");
        }
        else {
            raw.matchOutput.forEach((entry, index) => {
                errors.push(...validateOutputMatch(entry, `matchOutput[${index}]`));
            });
        }
    }
    if (errors.length > 0) {
        return {
            ok: false,
            errors,
        };
    }
    return { ok: true };
}
export function assertValidRule(raw) {
    const validation = validateRule(raw);
    if (!validation.ok) {
        throw new Error(`invalid rule:\n- ${validation.errors.join("\n- ")}`);
    }
}
