// tokenjuice/dist/hosts/openhands/index.js
import { mkdir, readFile, rename, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { compactBashResult } from "../../core/integrations/compact-bash-result.js";
import { buildTokenjuiceHookCommand, } from "../shared/host-command.js";
import { buildHookCommandDoctorFields } from "../shared/hook-command-doctor.js";
import { buildCompactedOutputContext, writeEmptyHookJsonLine, writeHookJsonLine } from "../shared/hook-output.js";
import { isRecord } from "../shared/hooks-json-file.js";
const TOKENJUICE_OPENHANDS_SUBCOMMAND = "openhands-post-tool-use";
const TOKENJUICE_OPENHANDS_FIX_COMMAND = "tokenjuice install openhands";
const TOKENJUICE_OPENHANDS_ADVISORY = "OpenHands support is beta and currently injects compacted context without suppressing the original tool result.";
function getProjectDir(options = {}) {
    return options.projectDir || process.env.OPENHANDS_PROJECT_DIR || process.cwd();
}
function getDefaultHooksPath(options = {}) {
    return join(getProjectDir(options), ".openhands", "hooks.json");
}
function sanitizeOpenHandsHooksConfig(raw) {
    if (!isRecord(raw)) {
        return { post_tool_use: [] };
    }
    return {
        ...raw,
        post_tool_use: Array.isArray(raw.post_tool_use) ? [...raw.post_tool_use] : [],
    };
}
async function readOpenHandsHooksConfig(hooksPath) {
    try {
        const rawText = await readFile(hooksPath, "utf8");
        return { config: sanitizeOpenHandsHooksConfig(JSON.parse(rawText)), exists: true };
    }
    catch (error) {
        if (error.code === "ENOENT") {
            return { config: { post_tool_use: [] }, exists: false };
        }
        throw error;
    }
}
async function loadOpenHandsHooksConfigWithBackup(hooksPath) {
    try {
        const rawText = await readFile(hooksPath, "utf8");
        const backupPath = `${hooksPath}.bak`;
        await writeFile(backupPath, rawText, "utf8");
        return { config: sanitizeOpenHandsHooksConfig(JSON.parse(rawText)), backupPath };
    }
    catch (error) {
        if (error.code === "ENOENT") {
            return { config: { post_tool_use: [] } };
        }
        throw error;
    }
}
async function writeOpenHandsHooksConfig(hooksPath, config) {
    await mkdir(dirname(hooksPath), { recursive: true });
    const tempPath = `${hooksPath}.tmp`;
    await writeFile(tempPath, `${JSON.stringify(config, null, 2)}\n`, "utf8");
    await rename(tempPath, hooksPath);
}
function isTokenjuiceOpenHandsHook(hook) {
    return isRecord(hook)
        && typeof hook.command === "string"
        && hook.command.includes(TOKENJUICE_OPENHANDS_SUBCOMMAND);
}
function findTokenjuiceOpenHandsHookCommand(config) {
    for (const group of config.post_tool_use) {
        if (!isRecord(group) || !Array.isArray(group.hooks)) {
            continue;
        }
        for (const hook of group.hooks) {
            if (isTokenjuiceOpenHandsHook(hook)) {
                return hook.command;
            }
        }
    }
    return undefined;
}
function removeTokenjuiceOpenHandsHooks(config) {
    let removed = 0;
    const retainedGroups = [];
    for (const group of config.post_tool_use) {
        if (!isRecord(group) || !Array.isArray(group.hooks)) {
            retainedGroups.push(group);
            continue;
        }
        const retainedHooks = group.hooks.filter((hook) => {
            const remove = isTokenjuiceOpenHandsHook(hook);
            if (remove) {
                removed += 1;
            }
            return !remove;
        });
        if (retainedHooks.length > 0) {
            retainedGroups.push({ ...group, hooks: retainedHooks });
        }
    }
    config.post_tool_use = retainedGroups;
    return removed;
}
function createOpenHandsHook(command) {
    return {
        matcher: "terminal",
        hooks: [
            {
                type: "command",
                command,
                timeout: 60,
            },
        ],
    };
}
export async function installOpenHandsHook(hooksPath, options = {}) {
    const resolvedHooksPath = hooksPath ?? getDefaultHooksPath(options);
    const { config, backupPath } = await loadOpenHandsHooksConfigWithBackup(resolvedHooksPath);
    const command = await buildTokenjuiceHookCommand(TOKENJUICE_OPENHANDS_SUBCOMMAND, "openhands", options);
    removeTokenjuiceOpenHandsHooks(config);
    config.post_tool_use = [...config.post_tool_use, createOpenHandsHook(command)];
    await writeOpenHandsHooksConfig(resolvedHooksPath, config);
    return {
        hooksPath: resolvedHooksPath,
        ...(backupPath ? { backupPath } : {}),
        command,
    };
}
export async function uninstallOpenHandsHook(hooksPath = getDefaultHooksPath()) {
    const { config } = await readOpenHandsHooksConfig(hooksPath);
    const removed = removeTokenjuiceOpenHandsHooks(config);
    if (removed > 0) {
        await writeOpenHandsHooksConfig(hooksPath, config);
    }
    return { hooksPath, removed };
}
export async function doctorOpenHandsHook(hooksPath, options = {}) {
    const resolvedHooksPath = hooksPath ?? getDefaultHooksPath(options);
    const expectedCommand = await buildTokenjuiceHookCommand(TOKENJUICE_OPENHANDS_SUBCOMMAND, "openhands", options);
    const { config, exists } = await readOpenHandsHooksConfig(resolvedHooksPath);
    const detectedCommand = findTokenjuiceOpenHandsHookCommand(config);
    return {
        hooksPath: resolvedHooksPath,
        ...(await buildHookCommandDoctorFields({
            expectedCommand,
            detectedCommand: exists ? detectedCommand : undefined,
            disabledIssue: "tokenjuice PostToolUse hook is not installed for OpenHands",
            hostLabel: "OpenHands",
            advisory: TOKENJUICE_OPENHANDS_ADVISORY,
            fixCommand: TOKENJUICE_OPENHANDS_FIX_COMMAND,
        })),
    };
}
function readStringField(record, keys) {
    for (const key of keys) {
        const value = record[key];
        if (typeof value === "string" && value.trim()) {
            return value;
        }
    }
    return undefined;
}
function readOpenHandsOutputText(response) {
    if (typeof response === "string") {
        return response;
    }
    if (!isRecord(response)) {
        return "";
    }
    return readStringField(response, ["output", "text", "content", "result", "llmContent"]) ?? "";
}
export async function runOpenHandsPostToolUseHook(rawText) {
    let payload;
    try {
        payload = JSON.parse(rawText);
    }
    catch {
        writeEmptyHookJsonLine();
        return 0;
    }
    if (payload.event_type !== "PostToolUse" || payload.tool_name !== "terminal") {
        writeEmptyHookJsonLine();
        return 0;
    }
    const toolInput = isRecord(payload.tool_input) ? payload.tool_input : undefined;
    const command = toolInput ? readStringField(toolInput, ["command", "cmd"]) : undefined;
    const visibleText = readOpenHandsOutputText(payload.tool_response);
    if (!command || !visibleText.trim()) {
        writeEmptyHookJsonLine();
        return 0;
    }
    try {
        const outcome = await compactBashResult({
            source: "openhands",
            command,
            visibleText,
            ...(typeof payload.working_dir === "string" && payload.working_dir.trim() ? { cwd: payload.working_dir } : {}),
            inspectionPolicy: "allow-safe-inventory",
            metadata: { source: "openhands-post-tool-use" },
            genericFallbackMinSavedChars: 120,
            genericFallbackMaxRatio: 0.75,
            skipGenericFallbackForCompoundCommands: true,
        });
        if (outcome.action === "keep") {
            writeEmptyHookJsonLine();
            return 0;
        }
        writeHookJsonLine({
            additionalContext: buildCompactedOutputContext(outcome.result.inlineText),
        });
        return 0;
    }
    catch {
        writeEmptyHookJsonLine();
        return 0;
    }
}
