// tokenjuice/dist/core/integrations/compact-bash-result.js
import { getInspectionCommandSkipReason, getSafeRepositoryInventorySourceArgv } from "../inventory-safety.js";
import { buildInspectionSummary } from "../reduce-inspection-summary.js";
import { reduceExecution } from "../reduce.js";
import { getCompactionSkipReason } from "./rewrite-policy.js";
function resolveInspectionPolicy(input) {
    if (input.inspectionPolicy) {
        return input.inspectionPolicy;
    }
    return input.skipInspectionCommands ? "skip-all" : "compact-all";
}
export function getOutputAwareInspectionSkipReason(policy, executionInput) {
    const command = executionInput.command ?? "";
    const rawText = executionInput.combinedText ?? "";
    const inspectionSkipReason = getInspectionCommandSkipReason(command, policy);
    if (inspectionSkipReason === "file-content-inspection-command"
        && buildInspectionSummary(executionInput, rawText)) {
        return null;
    }
    return inspectionSkipReason;
}
export async function compactBashResult(input) {
    const command = input.command.trim();
    if (!command) {
        return {
            action: "keep",
            reason: "unsupported",
            rawText: "",
            usedTrustedFullText: false,
        };
    }
    const rawText = input.trustedFullText ?? input.visibleText;
    const usedTrustedFullText = typeof input.trustedFullText === "string";
    if (!rawText.trim()) {
        return {
            action: "keep",
            reason: "empty-output",
            rawText,
            usedTrustedFullText,
        };
    }
    const safeInventoryArgv = getSafeRepositoryInventorySourceArgv(command);
    const executionInput = {
        toolName: "exec",
        command,
        combinedText: rawText,
        ...(safeInventoryArgv ? { argv: safeInventoryArgv } : {}),
        ...(typeof input.cwd === "string" && input.cwd.trim() ? { cwd: input.cwd } : {}),
        ...(typeof input.exitCode === "number" ? { exitCode: input.exitCode } : {}),
        ...(input.metadata ? { metadata: input.metadata } : {}),
    };
    const inspectionSkipReason = getOutputAwareInspectionSkipReason(resolveInspectionPolicy(input), executionInput);
    if (inspectionSkipReason) {
        return {
            action: "keep",
            reason: inspectionSkipReason,
            rawText,
            usedTrustedFullText,
        };
    }
    const options = {
        ...(typeof input.cwd === "string" && input.cwd.trim() ? { cwd: input.cwd } : {}),
        ...(typeof input.maxInlineChars === "number" ? { maxInlineChars: input.maxInlineChars } : {}),
        recordStats: true,
        ...(input.storeRaw ? { store: true } : {}),
    };
    const result = await reduceExecution(executionInput, options);
    const skipReason = getCompactionSkipReason(command, rawText, result, input);
    if (skipReason) {
        return {
            action: "keep",
            reason: skipReason,
            rawText,
            usedTrustedFullText,
            result,
        };
    }
    return {
        action: "rewrite",
        rawText,
        usedTrustedFullText,
        result,
    };
}
