// openclaw/dist/extensionAPI.js
import { b as resolveAgentDir, x as resolveAgentWorkspaceDir } from "./agent-scope-B6RIBoEj.js";
import { n as DEFAULT_MODEL, r as DEFAULT_PROVIDER } from "./defaults-Cbe87E7A.js";
import { i as resolveSessionFilePath, u as resolveStorePath } from "./paths-DUlscpp0.js";
import { t as loadSessionStore } from "./store-load-Dys5caP1.js";
import { i as saveSessionStore, o as updateSessionStore, s as updateSessionStoreEntry } from "./store-BDbj36M4.js";
import "./sessions-B8M_z4fr.js";
import { p as resolveThinkingDefault } from "./model-selection-CAAffjMN.js";
import { t as resolveAgentTimeoutMs } from "./timeout-B2er_ODN.js";
import { l as ensureAgentWorkspace } from "./workspace-Ba1XgL88.js";
import { n as resolveAgentIdentity } from "./identity-D9Py3mDy.js";
import { t as runEmbeddedPiAgent } from "./pi-embedded-CM_pfO4f.js";
//#region src/extensionAPI.ts
if (process.env.VITEST !== "true" && process.env.OPENCLAW_SUPPRESS_EXTENSION_API_WARNING !== "1") process.emitWarning("openclaw/extension-api is deprecated. Migrate to api.runtime.agent.* or focused openclaw/plugin-sdk/<subpath> imports. See https://docs.openclaw.ai/plugins/sdk-migration", {
	code: "OPENCLAW_EXTENSION_API_DEPRECATED",
	detail: "This compatibility bridge is temporary. Bundled plugins should use the injected plugin runtime instead of importing host-side agent helpers directly. Migration guide: https://docs.openclaw.ai/plugins/sdk-migration"
});
//#endregion
export { DEFAULT_MODEL, DEFAULT_PROVIDER, ensureAgentWorkspace, loadSessionStore, resolveAgentDir, resolveAgentIdentity, resolveAgentTimeoutMs, resolveAgentWorkspaceDir, resolveSessionFilePath, resolveStorePath, resolveThinkingDefault, runEmbeddedPiAgent, saveSessionStore, updateSessionStore, updateSessionStoreEntry };
