// openclaw/dist/slash-dispatch.runtime-LiT6wLYO.js
import { t as resolveConversationLabel$1 } from "./conversation-label-B1YMQZf8.js";
import { s as resolveChunkMode$1 } from "./chunk-Dhvlxa7H.js";
import { i as resolveAgentRoute$1 } from "./resolve-route-23mGh_7V.js";
import { t as finalizeInboundContext$1 } from "./inbound-context-BDVckYFC.js";
import { n as dispatchReplyWithDispatcher$1 } from "./provider-dispatcher-B730zRLd.js";
import "./routing-CFCE0Z1M.js";
import { t as resolveMarkdownTableMode$1 } from "./markdown-tables-CpQ0XGl5.js";
import "./reply-runtime-CVC35hLN.js";
import { t as recordInboundSessionMetaSafe$1 } from "./conversation-runtime-BiqjNzpw.js";
import "./markdown-table-runtime-C44wHHyv.js";
import { r as deliverSlackSlashReplies$1 } from "./replies-D1DqzAij.js";
//#region extensions/slack/src/monitor/slash-dispatch.runtime.ts
function resolveChunkMode(...args) {
	return resolveChunkMode$1(...args);
}
function finalizeInboundContext(...args) {
	return finalizeInboundContext$1(...args);
}
function dispatchReplyWithDispatcher(...args) {
	return dispatchReplyWithDispatcher$1(...args);
}
function resolveConversationLabel(...args) {
	return resolveConversationLabel$1(...args);
}
function recordInboundSessionMetaSafe(...args) {
	return recordInboundSessionMetaSafe$1(...args);
}
function resolveMarkdownTableMode(...args) {
	return resolveMarkdownTableMode$1(...args);
}
function resolveAgentRoute(...args) {
	return resolveAgentRoute$1(...args);
}
function deliverSlackSlashReplies(...args) {
	return deliverSlackSlashReplies$1(...args);
}
//#endregion
export { deliverSlackSlashReplies, dispatchReplyWithDispatcher, finalizeInboundContext, recordInboundSessionMetaSafe, resolveAgentRoute, resolveChunkMode, resolveConversationLabel, resolveMarkdownTableMode };
