// tokenjuice/dist/hosts/pi/index.js
import { access, mkdir, readFile, rename, rm, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { homedir } from "node:os";
import { fileURLToPath } from "node:url";
import packageJson from "../../../package.json" with { type: "json" };
const TOKENJUICE_PI_EXTENSION_COMMAND = "tj";
const LEGACY_PI_RUNTIME_BASENAME = "tokenjuice-runtime.js";
const TOKENJUICE_PI_FIX_COMMAND = "tokenjuice install pi";
function getPiAgentDir() {
    return process.env.PI_CODING_AGENT_DIR || join(homedir(), ".pi", "agent");
}
function getDefaultExtensionPath() {
    return join(getPiAgentDir(), "extensions", "tokenjuice.js");
}
function buildPiExtensionSource(runtimeSource, config) {
    return String.raw `/**
 * tokenjuice pi extension
 * generated by tokenjuice v${packageJson.version}
 */

${runtimeSource}

const config = ${JSON.stringify(config, null, 2)};

export default createTokenjuicePiExtension(config);
`;
}
async function resolvePiRuntimeEntryPoint() {
    const candidates = [
        fileURLToPath(new URL("./extension/runtime.ts", import.meta.url)),
        fileURLToPath(new URL("../../../src/hosts/pi/extension/runtime.ts", import.meta.url)),
    ];
    for (const candidate of candidates) {
        try {
            await access(candidate);
            return candidate;
        }
        catch {
            continue;
        }
    }
    throw new Error(`unable to resolve pi runtime source entrypoint from ${candidates.join(", ")}`);
}
async function bundlePiExtensionRuntimeSource() {
    let esbuild;
    try {
        esbuild = await import("esbuild");
    }
    catch (error) {
        throw new Error(`unable to load esbuild to bundle the pi runtime from source: ${String(error)}`);
    }
    const entryPoint = await resolvePiRuntimeEntryPoint();
    const result = await esbuild.build({
        entryPoints: [entryPoint],
        bundle: true,
        format: "esm",
        platform: "node",
        target: "node20",
        write: false,
    });
    const runtimeSource = result.outputFiles[0]?.text;
    if (!runtimeSource) {
        throw new Error("failed to bundle the pi runtime: esbuild produced no output");
    }
    return runtimeSource;
}
async function loadPiExtensionRuntimeSource(local = false) {
    if (local) {
        return await bundlePiExtensionRuntimeSource();
    }
    const runtimeAssetUrl = new URL("./extension/runtime.js", import.meta.url);
    try {
        return await readFile(runtimeAssetUrl, "utf8");
    }
    catch (error) {
        if (error.code !== "ENOENT") {
            throw error;
        }
    }
    return await bundlePiExtensionRuntimeSource();
}
async function writeAtomically(path, content) {
    const tempPath = `${path}.tmp`;
    await writeFile(tempPath, content, { encoding: "utf8", mode: 0o600 });
    await rename(tempPath, path);
}
async function removeLegacyRuntimeFile(extensionPath) {
    const legacyRuntimePath = join(dirname(extensionPath), LEGACY_PI_RUNTIME_BASENAME);
    try {
        await rm(legacyRuntimePath);
    }
    catch (error) {
        if (error.code !== "ENOENT") {
            throw error;
        }
    }
}
export async function installPiExtension(extensionPath = getDefaultExtensionPath(), options = {}) {
    const runtimeConfig = {
        extensionCommand: TOKENJUICE_PI_EXTENSION_COMMAND,
    };
    const runtimeSource = await loadPiExtensionRuntimeSource(Boolean(options.local));
    const extensionSource = buildPiExtensionSource(runtimeSource, runtimeConfig);
    await mkdir(dirname(extensionPath), { recursive: true });
    let backupPath;
    try {
        const existing = await readFile(extensionPath, "utf8");
        backupPath = `${extensionPath}.bak`;
        await writeFile(backupPath, existing, "utf8");
    }
    catch (error) {
        if (error.code !== "ENOENT") {
            throw error;
        }
    }
    await removeLegacyRuntimeFile(extensionPath);
    await writeAtomically(extensionPath, extensionSource);
    return {
        extensionPath,
        ...(backupPath ? { backupPath } : {}),
    };
}
export async function doctorPiExtension(extensionPath = getDefaultExtensionPath()) {
    let source;
    try {
        source = await readFile(extensionPath, "utf8");
    }
    catch (error) {
        if (error.code === "ENOENT") {
            return {
                extensionPath,
                status: "disabled",
                issues: [],
                fixCommand: TOKENJUICE_PI_FIX_COMMAND,
                checkedPaths: [],
                missingPaths: [],
            };
        }
        return {
            extensionPath,
            status: "broken",
            issues: [`tokenjuice pi extension could not be read: ${String(error)}`],
            fixCommand: TOKENJUICE_PI_FIX_COMMAND,
            checkedPaths: [extensionPath],
            missingPaths: [],
        };
    }
    const issues = [];
    if (!source.includes("generated by tokenjuice v") || !source.includes("createTokenjuicePiExtension")) {
        issues.push("installed pi extension does not look like a tokenjuice-generated extension");
    }
    if (source.includes(LEGACY_PI_RUNTIME_BASENAME)) {
        issues.push("installed pi extension still references the legacy sibling runtime");
    }
    return {
        extensionPath,
        status: issues.length > 0 ? "warn" : "ok",
        issues,
        fixCommand: TOKENJUICE_PI_FIX_COMMAND,
        checkedPaths: [extensionPath],
        missingPaths: [],
    };
}
